lib/spack/spack/environment.py

```diff
@@ -3,621 +3,866 @@
-    ``/usr/include`` and ``/usr/lib64``. If found we want to prepend
-    ``/usr/include`` to ``CPATH`` and ``/usr/lib64`` to ``MY_LIB64_PATH``.
+        # put the new user specs in the YAML
+        yaml_spec_list = config_dict(self.yaml).setdefault('specs', [])
+        yaml_spec_list[:] = [str(s) for s in self.user_specs]
 
-        .. code-block:: python
+        # if all that worked, write out the manifest file at the top level
+        with fs.write_tmp_and_move(self.manifest_path) as f:
+            _write_yaml(self.yaml, f)
 
-            # Set up the dictionary containing the inspection
-            inspections = {
-                'include': ['CPATH'],
-                'lib64': ['MY_LIB64_PATH']
-            }
+    def __enter__(self):
+        self._previous_active = active
+        activate(self)
+        return
 
-            # Get back the list of command needed to modify the environment
-            env = inspect_path('/usr', inspections)
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        deactivate()
+        if self._previous_active:
+            activate(self._previous_active)
 
-            # Eventually execute the commands
-            env.apply_modifications()
 
-    Returns:
-        instance of EnvironmentModifications containing the requested
-        modifications
-    """
-    if exclude is None:
-        exclude = lambda x: False
+def make_repo_path(root):
+    """Make a RepoPath from the repo subdirectories in an environment."""
+    path = spack.repo.RepoPath()
+
+    if os.path.isdir(root):
+        for repo_root in os.listdir(root):
+            repo_root = os.path.join(root, repo_root)
+
+            if not os.path.isdir(repo_root):
+                continue
+
+            repo = spack.repo.Repo(repo_root)
+            path.put_last(repo)
+
+    return path
 
-    env = EnvironmentModifications()
-    # Inspect the prefix to check for the existence of common directories
-    for relative_path, variables in inspections.items():
-        expected = os.path.join(root, relative_path)
 
-        if os.path.isdir(expected) and not exclude(expected):
-            for variable in variables:
-                env.prepend_path(variable, expected)
+def prepare_config_scope(env):
+    """Add env's scope to the global configuration search path."""
+    for scope in env.config_scopes():
+        spack.config.config.push_scope(scope)
 
-    return env
 
+def deactivate_config_scope(env):
+    """Remove any scopes from env from the global config path."""
+    for scope in env.config_scopes():
+        spack.config.config.remove_scope(scope.name)
 
-@contextlib.contextmanager
-def preserve_environment(*variables):
-    """Ensures that the value of the environment variables passed as
-    arguments is the same before entering to the context manager and after
-    exiting it.
 
-    Variables that are unset before entering the context manager will be
-    explicitly unset on exit.
+class EnvError(spack.error.SpackError):
```

Comment:
Can this be called `SpackEnvironmentError`?

Reply:
Done!